chain-tx-enclave-next/tx-query-next/app-runner/src/zmq_helper.rs

```diff
@@ -0,0 +1,189 @@
+use std::{
+    convert::TryInto,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    vec::IntoIter,
+};
+
+use async_zmq::{request, Message, MultipartIter, Request};
+use futures::ready;
+use tokio::{
+    io::{self, AsyncRead, AsyncWrite},
+    sync::Mutex,
+};
+
+/// Temporary bridging between the old code and EDP. In the long term, it may go away completely.
+///
+/// 1. Decryption requests may include the sealed payload in them (so no need to request them via zmq)
+/// 2. Encryption requests may be done via a direct attested secure channel (instead of this passing of sealed requests)
+pub struct ZmqHelper {
+    inner: Mutex<Inner>,
+}
+
+/// Inner struct to hold values in a `Mutex`
+///
+/// `Mutex` is there to enforce the synchronous order as zmq isn't thread safe
+struct Inner {
+    /// ZeroMQ socket
+    socket: Request<IntoIter<Message>, Message>,
+    /// State of ZeroMQ socket
+    state: SocketState,
+}
+
+impl ZmqHelper {
+    pub fn new(endpoint: &str) -> io::Result<Self> {
+        let socket = request(endpoint)
+            .map_err(into_io_err)?
+            .connect()
+            .map_err(into_io_err)?;
+
+        Ok(ZmqHelper {
+            inner: Mutex::new(Inner {
+                socket,
+                state: SocketState::Ready,
+            }),
+        })
+    }
+}
+
+/// State of ZeroMQ socket
+#[derive(Debug)]
+enum SocketState {
+    /// Denotes that a new request can be sent on ZeroMQ socket
+    Ready,
+    /// Denotes that a request is sent to ZeroMQ socket but response is not received yet
+    RequestSent,
+    /// Denotes that response to a previously sent request is received
+    ResponseReceived(Message),
+}
+
+impl SocketState {
+    fn is_ready(&self) -> bool {
+        matches!(self, SocketState::Ready)
+    }
+
+    fn is_request_sent(&self) -> bool {
+        matches!(self, SocketState::RequestSent)
+    }
+
+    fn is_response_received(&self) -> bool {
+        matches!(self, SocketState::ResponseReceived(_))
+    }
+
+    fn unwrap(self) -> Message {
+        match self {
+            SocketState::ResponseReceived(message) => message,
+            _ => panic!("Cannot fetch response. Haven't received it yet."),
+        }
+    }
+}
+
+impl AsyncRead for ZmqHelper {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut inner_future = Box::pin(self.inner.lock());
+        let mut inner = ready!(inner_future.as_mut().poll(cx));
+
+        if inner.state.is_ready() {
+            log::debug!("Unable to receive message. Previous request wasn't sent");
```

Comment:
Should this be an error or panic?

Reply:
I don't think we should panic here. I'll add a way to wake all the tasks waiting for state change.

Reply:
I added a waking mechanism for tasks waiting for ready state. At other places, it is ok to panic because those are logically unreachable.